elastic/integrations
# [Logstash] Add CEL native data collection, and dashboards for Pipeline and Pipeline Details

## Proposed commit message

This PR follows on from #7704, and adds data collection to the integration, as well as standard Dashboards for Logstash Pipelines.

The short term goal for this is to enable Logstash to be monitored with Serverless Elasticsearch instances, which do not provide a Stack Monitoring UI, with the long term goal of iterating on this to replace the existing Stack Monitoring UI.

This PR includes 3 new data streams - `node_cel`, `pipelines` and `plugins`, with data collected from the Logstash Metrics API using the CEL input, intended to replace the existing metricbeat collection. 

This give us more flexibility going forward, and should improve our velocity to add new features to the Logstash dashboards straight from the API, without needing to update several code bases.

The CEL data streams will create a single document per node for the `node_cel` data stream, a document per pipeline per node for the `pipeline` data stream, and a document per plugin per pipeline per node for the `plugins` data stream. For this reason, we default to a lower sending cadence for the `plugins` data stream than for the other. 

Note that the existing data collection from metricbeat defaults to sending large documents, including significant data for the entire pipeline graph every 10 seconds. 

This integration also includes ingest pipelines to clean up the data from the Logstash flow metrics API, to remove entries we don't necessarily need at this stage, and to remove the string representation of "Infinity".

When installing a Logstash integration, users will be presented with an option to collect Metrics for Stack Monitoring or for Technical Preview

<img width="698" alt="Screenshot 2023-10-04 at 3 50 30 PM" src="https://github.com/elastic/integrations/assets/4061337/15e92e5e-c4fb-4813-862f-b1cd655fc4ce">

(Note: If there was a way to have an either/or option in the UI, that would be very helpful)

The dashboards are described in the _Screenshots_ section.



## Checklist



## How to test this PR locally


* Install [elastic-package](https://github.com/elastic/elastic-package)
* Checkout this PR
* Navigate to the `integrations/packages/logstash` directory
* Run `elastic-package build` to build the package
* Run `elastic-package stack up` to bring up the Elastic stack
* Start an instance of Logstash
* Install the `Logstash` integration
* Uncheck the `Metrics (Stack Monitoring)` option
* Check the `Metrics (Technical Preview)` option
* Click 'Save and Continue' and follow the instructions to enroll the agent

## Related issues

Follows on from #7704

- 

## Screenshots

Integration Selection Page


<img width="812" alt="Screenshot 2023-10-04 at 3 56 22 PM" src="https://github.com/elastic/integrations/assets/4061337/f304f47b-b63f-4b52-b5cf-efc302117bc3">


### Pipelines Dashboard

This Dashboard provides an overview of running pipelines, with a table containing details of pipelines running, which can be drilled down into a "details view"

#### Proposed Page

<img width="1770" alt="Screenshot 2023-10-04 at 4 08 37 PM" src="https://github.com/elastic/integrations/assets/4061337/0195d157-fc20-4f2f-8f23-2cb2a1c4e11f">


This is intended to replace the existing pipelines page in the Stack monitoring solution

##### Existing Page
<img width="1765" alt="Screenshot 2023-10-04 at 4 08 15 PM" src="https://github.com/elastic/integrations/assets/4061337/0ab3f36b-fb27-4a04-b1d4-a9d4872b1fb7">


The page includes additional details, including Persistent Queue usage per pipeline, and more visibility into which pipelines may be blocked


### Pipeline Details Dashboard

The Pipeline Details dashboard takes a different approach to the pipeline details view in stack monitoring

The page includes some general pipeline information, showing the rate of events flowing through the pipelines and state of persistent queues, broken down per node running the pipeline:

#### Proposed Page

##### General Information

<img width="1775" alt="Screenshot 2023-10-04 at 4 26 05 PM" src="https://github.com/elastic/integrations/assets/4061337/4675b0f1-9652-4535-bd42-874a3d45630c">

It also includes information for each plugin type: 

##### Inputs

<img width="1478" alt="Screenshot 2023-10-04 at 4 27 53 PM" src="https://github.com/elastic/integrations/assets/4061337/a3bcbfee-27c0-4ee6-813a-2598961f97cb">

##### Filters

<img width="1480" alt="Screenshot 2023-10-04 at 4 28 50 PM" src="https://github.com/elastic/integrations/assets/4061337/ba501884-3b2d-4855-a200-c2aa5f61b4c9">

###### Outputs

<img width="1478" alt="Screenshot 2023-10-04 at 4 29 49 PM" src="https://github.com/elastic/integrations/assets/4061337/2af4fe5f-ce7e-4cdd-b07c-79db1ea74f5f">

###### Codecs

<img width="1477" alt="Screenshot 2023-10-04 at 4 30 21 PM" src="https://github.com/elastic/integrations/assets/4061337/f2cf75d0-26ff-4f5d-b16a-3fbba086be4e">

This has benefits and drawbacks compared with the existing approach:

The benefits are:
* More information and more rich drill downs to help track down slow performing plugins
* Can track PQ usage on a per-pipeline and per-node basis

Drawbacks:
* The lack of visual representation may make it more difficult to figure out which plugin is at fault when multiple plugins of the same type are at play
  * Note - this can be mitigated by users including an id in their plugin definition
  * We will be looking to add more information (originating file + line number of plugin) in a subsequent follow-up PR.
 
#### Existing Dashboard

The existing dashboard contains a representation of the pipeline graph, which we are not easily able to show using standard Kibana Dashboards

<img width="346" alt="Screenshot 2023-10-04 at 4 32 50 PM" src="https://github.com/elastic/integrations/assets/4061337/b067d37a-9449-415d-af8e-9c29879a200b">

And a small pop-up drilldown for each plugin

<img width="439" alt="Screenshot 2023-10-04 at 4 36 56 PM" src="https://github.com/elastic/integrations/assets/4061337/0b683ac4-3e6c-4d35-8944-9402ac562c90">



Note: This will require a follow-up PR to enable for serverless - the existing integration has some compatibility issues that I would like to fold into a subsequent PR
